test-data/unit/check-classes.test

```diff
@@ -4223,3 +4224,35 @@ class C(Any):
         reveal_type(self.bar().__name__) # E: Revealed type is 'builtins.str'
 [builtins fixtures/type.pyi]
 [out]
+
+[case testClassDecoratorIsTypeChecked]
+from typing import Callable, Type
+def decorate(x: int) -> Callable[[type], type]:  # N: "decorate" defined here
+    ...
+def decorate_forward_ref() -> Callable[[Type[A]], Type[A]]:
+    ...
+@decorate(y=17)    # E: Unexpected keyword argument "y" for "decorate"
+@decorate()        # E: Too few arguments for "decorate"
+@decorate(22, 25)  # E: Too many arguments for "decorate"
+@decorate_forward_ref()
+@decorate(11)
+class A: pass
+
+@decorate  # E: Argument 1 to "decorate" has incompatible type "Type[A2]"; expected "int"
+class A2: pass
+
+[case testClassDecoratorIncorrect]
+def not_a_class_decorator(x: int) -> int: ...
+@not_a_class_decorator(7)  # E: "int" not callable
```

Comment:
The error message is quite obscure. I would somehow highlight that a class decorator must be a function that accepts a type. If this is too hard, then leave a TODO about this in code.

Reply:
Left a TODO. Only thing I can figure out is collect all the errors with `self.msg.copy()` then modify all the messages to say what we want.  Or check the count of errors before and after and add another one if it has increased.